queryable-spec.ts

```diff
@@ -0,0 +1,424 @@
+ * it must match each produced quad, according to the Quad.equals semantics. 
+ * When a Term parameter is a Variable, or it is undefined, it acts as a 
+ * wildcard, and can match with any Term.
+ * 
+ * NOTES:
+ * - When matching with graph set to undefined or null it MUST match all the
+ *   graphs (sometimes called the union graph). To match only the default graph
+ *   set graph to a DefaultGraph.
+ * - When an Expression parameter is defined, the complete quad stream is
+ *   filtered according to this expression. When it is undefined, no filter is
+ *   applied.
+ * 
+ * If parameters of type Variable with an equal variable name are in place,
+ * then the corresponding quad components in the resulting quad stream MUST be
+ * equal.
+ * Expression's MAY contain Variable Term's. If their variable names are equal
+ * to Variable's in the given quad pattern, then the Expression MUST be 
+ * instantiated for each variable's binding in the resulting quad stream when
+ * applying the Expression filter.
+ */
+interface FilterableSource {
+  matchExpression(
+    subject?: RDF.Term,
+    predicate?: RDF.Term,
+    obj?: RDF.Term,
+    graph?: RDF.Term,
+    expression?: Expression,
+    opts?: { 
+      limit?: number; 
+      offset?: number; 
+    },
+  ): FilterableResult;
+};
+
+
+/******************************************************************************
+                              QUERYABLE SOURCE
+ *****************************************************************************/
+
+/*
+ * Map-like representation of Bindings as using plain objects could lead
+ * to collisions between variable names and object properties. Support for
+ * immutability is required (but implementations are free to be mutable) which
+ * determines the return value of the set() and delete() methods to be an 
+ * instance of Bindings (potentially a different one).
+ */ 
+interface Bindings extends Iterable<[RDF.Variable, RDF.Term]> {
+  type: 'bindings';
+  has: (key: RDF.Variable) => boolean;
+  get: (key: RDF.Variable) => RDF.Term | undefined;
+  keys: () => Iterator<RDF.Variable>;
+  values: () => Iterator<RDF.Term>;
+  entries: () => Iterator<[RDF.Variable, RDF.Term]>;
+  forEach: (fn: (value: RDF.Term, key: RDF.Variable) => any) => void;
+  size: number;
+  [Symbol.iterator]: () => Iterator<[RDF.Variable, RDF.Term]>;
+}
+
+/*
+ * Bindings objects are created using a dedicated factory, keeping in line
+ * with DataFactory.quad(). This also helps with facilitating support for
+ * immutability. Basic helper methods must also be provided for the most 
+ * common manipulations of bindings objects.
+ */
+interface BindingsFactory {
+  bindings: (entries?: [RDF.Variable, RDF.Term][]) => Bindings;
+  filter: (bindings: Bindings, fn: (value: RDF.Term, key: RDF.Variable) => boolean) => Bindings;
+  map: (bindings: Bindings, fn: (value: RDF.Term, key: RDF.Variable) => RDF.Term) => Bindings;
+  merge: (left: Bindings, right: Bindings) => Bindings;
```

Comment:
If a join conflict occurs, do we want this to return `undefined`?
(I would in any case document the merge semantics as comment)

Reply:
Ah yes, I forgot to make the default behavior of `merge` explicit while refactoring. Yes, I think returning `undefined` would be a good move.